Terraform crash planning a diff between Null and empty Sets

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to Terraform Cloud/Enterprise, please contact tf-cloud@hashicorp.support.
If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://registry.terraform.io/browse/providers.

To fix problems, we need clear reproduction cases - we need to be able to see it happen locally. A reproduction case is ideally something a Terraform Core engineer can git-clone or copy-paste and run immediately, without inventing any details or context. 

* A short example can be directly copy-pasteable; longer examples should be in separate git repositories, especially if multiple files are needed
* Please include all needed context. For example, if you figured out that an expression can cause a crash, put the expression in a variable definition or a resource
* Set defaults on (or omit) any variables. The person reproducing it should not need to invent variable settings
* If multiple steps are required, such as running terraform twice, consider scripting it in a simple shell script. Providing a script can be easier than explaining what changes to make to the config between runs.
* Omit any unneeded complexity: remove variables, conditional statements, functions, modules, providers, and resources that are not needed to trigger the bug
* When possible, use the [null resource](https://www.terraform.io/docs/providers/null/resource.html) provider rather than a real provider in order to minimize external dependencies. We know this isn't always feasible. The Terraform Core team doesn't have deep domain knowledge in every provider, or access to every cloud platform for reproduction cases.

-->

### Terraform Version
<!---
Run `terraform version` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```console
% terraform version
Terraform v1.0.4
on darwin_amd64
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For Terraform configs larger than a few resources, or that involve multiple files, please make a GitHub repository that we can clone, rather than copy-pasting multiple files in here. For security, you can also encrypt the files using our GPG public key at https://www.hashicorp.com/security.
-->

```terraform
resource "aws_kms_key" "test" {
  key_policy = jsonencode({
    Id = "kms-tf-1"
    Statement = [
      {
        Action = "kms:*"
        Effect = "Allow"
        Principal = {
          AWS = "*"
        }

        Resource = "*"
        Sid      = "Enable IAM User Permissions"
      },
    ]
    Version = "2012-10-17"
  })

  pending_window_in_days = 8
}
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

```
% terraform plan 
╷
│ Warning: Provider development overrides are in effect
│ 
│ The following provider development overrides are set in the CLI configuration:
│  - hashicorp/aws in /Users/ewbankkit/go/bin
│  - hashicorp/aws-xxxxxxxx in /Users/ewbankkit/go/bin
│ 
│ The behavior may therefore not match any released version of the provider and applying changes may cause the state to become incompatible with published releases.
╵
aws_kms_key.test: Refreshing state... [id=2ce9555e-fa82-4510-88dd-e304e519b2b8]
panic: must not call SetVal with empty slice

goroutine 15 [running]:
github.com/zclconf/go-cty/cty.SetVal(0x47ee998, 0x0, 0x0, 0x47ee998, 0x0, 0x0, 0x0)
	/Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.9.0/cty/value_init.go:267 +0x865
github.com/hashicorp/terraform/internal/command/format.(*blockBodyDiffPrinter).writeNestedAttrDiff(0xc000e712a8, 0xc000174e88, 0x4, 0xc000dc6320, 0x3749130, 0xc0005c6e70, 0x0, 0x0, 0x3749130, 0xc001105080, ...)
	/Users/distiller/project/project/internal/command/format/diff.go:615 +0x246a
github.com/hashicorp/terraform/internal/command/format.(*blockBodyDiffPrinter).writeAttrDiff(0xc000adf2a8, 0xc000174e88, 0x4, 0xc000e04280, 0x3749130, 0xc0005c6e70, 0x0, 0x0, 0x3749130, 0xc001105080, ...)
	/Users/distiller/project/project/internal/command/format/diff.go:473 +0x7a7
github.com/hashicorp/terraform/internal/command/format.(*blockBodyDiffPrinter).writeAttrsDiff(0xc000e712a8, 0xc000f0bb00, 0x37490f8, 0xc0011053e0, 0x2f418e0, 0xc000e20ab0, 0x37490f8, 0xc001105770, 0x2f418e0, 0xc000e20bd0, ...)
	/Users/distiller/project/project/internal/command/format/diff.go:434 +0x685
github.com/hashicorp/terraform/internal/command/format.(*blockBodyDiffPrinter).writeBlockBodyDiff(0xc000e712a8, 0xc000f0bbf0, 0x37490f8, 0xc0011053e0, 0x2f418e0, 0xc000e20ab0, 0x37490f8, 0xc001105770, 0x2f418e0, 0xc000e20bd0, ...)
	/Users/distiller/project/project/internal/command/format/diff.go:358 +0x12f
github.com/hashicorp/terraform/internal/command/format.ResourceInstanceDrift(0x0, 0x0, 0x0, 0x4d, 0xc00072da40, 0xb, 0xc00072da4c, 0x4, 0x0, 0x0, ...)
	/Users/distiller/project/project/internal/command/format/diff.go:273 +0x6c5
github.com/hashicorp/terraform/internal/command/views.renderChangesDetectedByRefresh(0xc0005a3ae0, 0xc0005a3ae8, 0xc0002768b0, 0xc000612840, 0xc0000aac78)
	/Users/distiller/project/project/internal/command/views/plan.go:385 +0x5d4
github.com/hashicorp/terraform/internal/command/views.renderPlan(0xc0002e7d60, 0xc0002768b0, 0xc000612840)
	/Users/distiller/project/project/internal/command/views/plan.go:100 +0x70
github.com/hashicorp/terraform/internal/command/views.(*OperationHuman).Plan(0xc000797b90, 0xc0002e7d60, 0xc0002768b0)
	/Users/distiller/project/project/internal/command/views/operation.go:92 +0x45
github.com/hashicorp/terraform/internal/backend/local.(*Local).opPlan(0xc0002e72c0, 0x3747418, 0xc0003d4340, 0x3747418, 0xc0003d4380, 0xc00079afc0, 0xc0003d4300)
	/Users/distiller/project/project/internal/backend/local/backend_plan.go:147 +0x8fc
github.com/hashicorp/terraform/internal/backend/local.(*Local).Operation.func1(0xc000797cf0, 0xc000797d00, 0xc000797d10, 0xc0002e72c0, 0xc000797ce0, 0x3747418, 0xc0003d4340, 0x3747418, 0xc0003d4380, 0xc00079afc0, ...)
	/Users/distiller/project/project/internal/backend/local/backend.go:321 +0xe8
created by github.com/hashicorp/terraform/internal/backend/local.(*Local).Operation
	/Users/distiller/project/project/internal/backend/local/backend.go:315 +0x2f2




!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.209140014.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

SECURITY WARNING: the "crash.209140014.log" file that was created may contain 
sensitive information that must be redacted before it is safe to share 
on the issue tracker.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

[crash.209140014.log](https://github.com/hashicorp/terraform/files/6993556/crash.209140014.log)

### Expected Behavior
<!--
What should have happened?
-->

### Actual Behavior
<!--
What actually happened?
-->

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->

**N.B.** This is not the [usual AWS Provider](https://github.com/hashicorp/terraform-provider-aws) but is an under-development version using https://github.com/hashicorp/terraform-plugin-framework (and hence protocol v6).

The resource's schema includes a `Set` of `Object`s (the resource's `tags`) which I have not specified in configuration but _are_ returned from AWS as an empty array :`"tags": [],`.

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

--- a/internal/command/format/diff.go
+++ b/internal/command/format/diff.go
@@ -471,8 +471,28 @@ func (p *blockBodyDiffPrinter) writeAttrDiff(name string, attrS *configschema.At
 	}
 
 	if attrS.NestedType != nil {
-		p.writeNestedAttrDiff(name, attrS.NestedType, old, new, nameLen, indent, path, action, showJustNew)
-		return false
+		renderNested := true
+
+		// If the collection values are empty or null, we render them as single attributes
+		switch attrS.NestedType.Nesting {
+		case configschema.NestingList, configschema.NestingSet, configschema.NestingMap:
+			var oldLen, newLen int
+			if !old.IsNull() && old.IsKnown() {
+				oldLen = old.LengthInt()
+			}
+			if !new.IsNull() && new.IsKnown() {
+				newLen = new.LengthInt()
+			}
+
+			if oldLen+newLen == 0 {
+				renderNested = false
+			}
+		}
+
+		if renderNested {
+			p.writeNestedAttrDiff(name, attrS.NestedType, old, new, nameLen, indent, path, action, showJustNew)
+			return false
+		}
 	}
 
 	p.buf.WriteString("\n")
@@ -613,6 +633,7 @@ func (p *blockBodyDiffPrinter) writeNestedAttrDiff(
 		allItems := make([]cty.Value, 0, len(oldItems)+len(newItems))
 		allItems = append(allItems, oldItems...)
 		allItems = append(allItems, newItems...)
+
 		all := cty.SetVal(allItems)
 
 		p.buf.WriteString(" = [")

--- a/internal/command/format/diff.go
+++ b/internal/command/format/diff.go
@@ -646,11 +646,11 @@ func (p *blockBodyDiffPrinter) writeNestedAttrDiff(
 			case !val.IsKnown():
 				action = plans.Update
 				newValue = val
-			case !old.HasElement(val).True():
+			case old.IsNull() || !old.HasElement(val).True():
 				action = plans.Create
 				oldValue = cty.NullVal(val.Type())
 				newValue = val
-			case !new.HasElement(val).True():
+			case new.IsNull() || !new.HasElement(val).True():
 				action = plans.Delete
 				oldValue = val
 				newValue = cty.NullVal(val.Type())

--- a/internal/command/format/diff_test.go
+++ b/internal/command/format/diff_test.go
@@ -2898,6 +2898,62 @@ func TestResourceChange_nestedSet(t *testing.T) {
     }
 `,
 		},
+		"in-place update - null insertion": {
+			Action: plans.Update,
+			Mode:   addrs.ManagedResourceMode,
+			Before: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.StringVal("i-02ae66f368e8518a9"),
+				"ami": cty.StringVal("ami-BEFORE"),
+				"disks": cty.NullVal(cty.Set(cty.Object(map[string]cty.Type{
+					"mount_point": cty.String,
+					"size":        cty.String,
+				}))),
+				"root_block_device": cty.SetVal([]cty.Value{
+					cty.ObjectVal(map[string]cty.Value{
+						"volume_type": cty.StringVal("gp2"),
+						"new_field":   cty.NullVal(cty.String),
+					}),
+				}),
+			}),
+			After: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.StringVal("i-02ae66f368e8518a9"),
+				"ami": cty.StringVal("ami-AFTER"),
+				"disks": cty.SetVal([]cty.Value{
+					cty.ObjectVal(map[string]cty.Value{
+						"mount_point": cty.StringVal("/var/diska"),
+						"size":        cty.StringVal("50GB"),
+					}),
+				}),
+				"root_block_device": cty.SetVal([]cty.Value{
+					cty.ObjectVal(map[string]cty.Value{
+						"volume_type": cty.StringVal("gp2"),
+						"new_field":   cty.StringVal("new_value"),
+					}),
+				}),
+			}),
+			RequiredReplace: cty.NewPathSet(),
+			Schema:          testSchemaPlus(configschema.NestingSet),
+			ExpectedOutput: `  # test_instance.example will be updated in-place
+  ~ resource "test_instance" "example" {
+      ~ ami   = "ami-BEFORE" -> "ami-AFTER"
+      + disks = [
+          + {
+            + mount_point = "/var/diska"
+            + size        = "50GB"
+          },
+        ]
+        id    = "i-02ae66f368e8518a9"
+
+      + root_block_device {
+          + new_field   = "new_value"
+          + volume_type = "gp2"
+        }
+      - root_block_device {
+          - volume_type = "gp2" -> null
+        }
+    }
+`,
+		},
 	}
 	runTestCases(t, testCases)
 }

--- a/internal/command/format/diff_test.go
+++ b/internal/command/format/diff_test.go
@@ -2863,6 +2863,41 @@ func TestResourceChange_nestedSet(t *testing.T) {
     }
 `,
 		},
+		"in-place update - empty nested sets": {
+			Action: plans.Update,
+			Mode:   addrs.ManagedResourceMode,
+			Before: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.StringVal("i-02ae66f368e8518a9"),
+				"ami": cty.StringVal("ami-BEFORE"),
+				"disks": cty.NullVal(cty.Set(cty.Object(map[string]cty.Type{
+					"mount_point": cty.String,
+					"size":        cty.String,
+				}))),
+				"root_block_device": cty.SetValEmpty(cty.Object(map[string]cty.Type{
+					"volume_type": cty.String,
+				})),
+			}),
+			After: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.StringVal("i-02ae66f368e8518a9"),
+				"ami": cty.StringVal("ami-AFTER"),
+				"disks": cty.SetValEmpty(cty.Object(map[string]cty.Type{
+					"mount_point": cty.String,
+					"size":        cty.String,
+				})),
+				"root_block_device": cty.SetValEmpty(cty.Object(map[string]cty.Type{
+					"volume_type": cty.String,
+				})),
+			}),
+			RequiredReplace: cty.NewPathSet(),
+			Schema:          testSchema(configschema.NestingSet),
+			ExpectedOutput: `  # test_instance.example will be updated in-place
+  ~ resource "test_instance" "example" {
+      ~ ami   = "ami-BEFORE" -> "ami-AFTER"
+      + disks = []
+        id    = "i-02ae66f368e8518a9"
+    }
+`,
+		},
 	}
 	runTestCases(t, testCases)
 }

